### PR TITLE
Refactor code

### PR DESCRIPTION
**Pull request by Bachibouzouk**
The idea is to refactor the code in order to facilitate testing of individual parts of the algorithm and to ease new developers into the understanding of what is being done where.

To facilitate the reviewers' job I took some time so that if you review commit by commit you should have a good understanding of each steps of the refactoring process (just looking at the final diff in that case might be frightening as most of the file changed)

[RAMP-project:development](https://github.com/RAMP-project/RAMP/tree/development) should maybe rebased on top of [RAMP-project:master](https://github.com/RAMP-project/RAMP/tree/master) to include the changes introduced by @mohammadamint in 5fe15064c630fb6d4e46ba24f831521e435a1e0f

### CODE REVIEW

**Review by FLomb — DISMISSED** [review]

File: ramp/core/core.py [hunk 1]
```diff
@@ -556,13 +554,28 @@ def assign_random_cycles(self):
...
                 )) #randomise also the fixed cycle
+
                 if self.fixed_cycle >= 3:
                     p_31 = random_variation(var=self.thermal_p_var, norm=self.p_31) #randomly variates the power of thermal apps, otherwise variability is 0
                     p_32 = random_variation(var=self.thermal_p_var, norm=self.p_32) #randomly variates the power of thermal apps, otherwise variability is 0
-                    self.random_cycle3 = np.concatenate((
-                        (p_31 * np.ones(int(random_variation(var=-self.r_c3, norm=self.t_31)))),
-                        (p_32 * np.ones(int(random_variation(var=-self.r_c3, norm=self.t_32))))
-                    )) #randomise also the fixed cycle
+                    self.random_cycle1 = random.choice([np.concatenate(((np.ones(
```
Comment (FLomb) [inline ramp/core/core.py]: I do not understand the following: why do we define 3 random cycles if self.fixed_cycle >= 3 but we define only random_cycle2 (and not random_cycle1) when self.fixed_cycle >= 2 ?

File: ramp/input_files/input_file_4.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
```
Comment (FLomb) [inline ramp/input_files/input_file_4.py:1]: I understand that this new input file is meant to test the spike problem. But do we need it as a permanent file addition if the issue has been debugged already?

**Comment by Bachibouzouk** [comment]
@FLomb - I see you are reviewing this PR, let me know if I can help in any way :)

File: ramp/core/core.py [hunk 3]
```diff
@@ -533,3 +533,20 @@ def switch_on(self, rand_window_1, rand_window_2, rand_window_3):
...
+                return None # this will trigger a continue in the loop
+            else:
+                upper_limit = next_switch[0] - switch_on  # if there are no other options to reach the total time of use, empty spaces are filled without minimum cycle restrictions until reaching the limit
+        else:
+            upper_limit = min(rand_time, rand_window[1] - switch_on)  # if there are no other switch-on events after the current one, the upper duration limit is set this way
+        if upper_limit >= self.func_cycle:
+            indexes = np.arange(switch_on, switch_on + (int(random.uniform(self.func_cycle, upper_limit))))
+        else:
+            indexes = np.arange(switch_on, switch_on + upper_limit)
+        return indexes
```
Comment (FLomb) [inline ramp/core/core.py]: I see that GitHub here complains about "no new line at end of file"

File: ramp/core/utils.py [hunk 4]
```diff
@@ -72,7 +72,6 @@ def read_input_file(filename):
                                 f"The provided power timeseries of the appliance '{appliance_name}' of user '{user_name}' in '{filename}' does not contain 365 values as expected\n{POSSIBLE_FORMATS}"
                             )
                         )
-
```
Comment (FLomb) [inline ramp/core/utils.py:75]: All I see in this commit is an empty line being deleted. Is that normal?

**Review by FLomb — COMMENTED** [review]
Thanks for this PR @Bachibouzouk! It looks great. I have just a few comments about things I would like you to double-check. In general, though, it looks like some commits do not exactly show what the title suggests (e.g., see my comment about the 'xls parser' commit). Can you double-check if all commits are as expected and that all is in order?

**Comment by FLomb** [comment]
> @FLomb - I see you are reviewing this PR, let me know if I can help in any way :)

Thanks, I have just finished and provided my comments above!

File: ramp/input_files/input_file_4.py (hunk 2 above)
Comment (Bachibouzouk) [reply ramp/input_files/input_file_4.py:1]: I think this is a mistake commit, I will remove it

File: ramp/core/core.py (hunk 1 above)
Comment (Bachibouzouk) [reply ramp/core/core.py]: I will look again into it before replying :)

**Comment by Bachibouzouk** [comment]
> I have indeed rebased the development branch to include the latest commit by @mohammadamint

@FLomb - I think you merged the `development` branch into itself, if you would have rebased it, one would not see the merge commit anymore

File: ramp/input_files/input_file_4.py (hunk 2 above)
Comment (Bachibouzouk) [reply ramp/input_files/input_file_4.py]: I removed it

File: ramp/core/utils.py (hunk 4 above)
Comment (Bachibouzouk) [reply ramp/core/utils.py:75]: I removed this commit

File: ramp/core/core.py (hunk 3 above)
Comment (Bachibouzouk) [reply ramp/core/core.py]: I removed this file anyway now :)

**Comment by FLomb** [comment]
> > I have indeed rebased the development branch to include the latest commit by @mohammadamint
> 
> @FLomb - I think you merged the `development` branch into itself, if you would have rebased it, one would not see the merge commit anymore. The last commit message on the branch reads: "Merge branch 'development' of https://github.com/RAMP-project/RAMP into development". What would need to be done is `git rebase -i master` while on the `development` branch, then `git push -f`

And yet I do see the contributions by @mohammadamint into the development branch (here: https://github.com/RAMP-project/RAMP/blob/development/.gitignore). And if I compare the two branches, the master is not seen as having any new change compared to the development. Not sure what happened, but it looks good to me.

File: ramp/core/core.py (hunk 1 above)
Comment (FLomb) [reply ramp/core/core.py]: You're right; there is also a comment in the original code that explains why it is so

**Comment by Bachibouzouk** [comment]
> Not sure what happened, but it looks good to me.

I mentionned it in the first post of this PR :

[RAMP-project:development](https://github.com/RAMP-project/RAMP/tree/development) should maybe rebased on top of [RAMP-project:master](https://github.com/RAMP-project/RAMP/tree/master) to include the changes introduced by @mohammadamint in https://github.com/RAMP-project/RAMP/commit/5fe15064c630fb6d4e46ba24f831521e435a1e0f

**Comment by FLomb** [comment]
> > Not sure what happened, but it looks good to me.
> 
> I mentionned it in the first post of this PR :
> 
> [RAMP-project:development](https://github.com/RAMP-project/RAMP/tree/development) should maybe rebased on top of [RAMP-project:master](https://github.com/RAMP-project/RAMP/tree/master) to include the changes introduced by @mohammadamint in [5fe1506](https://github.com/RAMP-project/RAMP/commit/5fe15064c630fb6d4e46ba24f831521e435a1e0f)

What I mean is that [5fe1506](https://github.com/RAMP-project/RAMP/commit/5fe15064c630fb6d4e46ba24f831521e435a1e0f) is already featured in both the master and the development branches, as it should. So there should be no need for rebasing. Did you mean that you did rebase yourself and that's why the commit now appears in both branches?

**Comment by Bachibouzouk** [comment]
> Did you mean that the reason the commit now appears in both branches is that you rebased it?

I don't think I had the rights to do this, maybe @mohammadamint did update `development` ? In any case if it is as it should it is for the best :), could we then merge? or do we need to make some more changes/review addition to the documentation?

**Review by FLomb — APPROVED** [review]

**Comment by FLomb** [comment]
I have just merged this PR into the development branch. People should now try and test it based on whatever they normally use RAMP for. @mohammadamint could you give us feedback since you were planning on using RAMP anyway in the coming weeks? I will also ask the same of other users and will test the code myself.

**Comment by FLomb** [comment]
@Bachibouzouk I have just noticed that the changelog hasn't really been updated in your PR. Could you please take care of that?

**Comment by Bachibouzouk** [comment]
> @Bachibouzouk I have just noticed that the changelog hasn't really been updated in your PR

I did not think you would accept and merge so quickly ^^ I was planning to add the changelog for the accepted changes. I can do a PR to amend the changelog :)

**Comment by Bachibouzouk** [comment]
The changelog is awaiting for review/approval and merge https://github.com/RAMP-project/RAMP/pull/40

**Pull request merged**